{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MGnify\n",
    "https://www.ebi.ac.uk/metagenomics/\n",
    "\n",
    "\n",
    "\n",
    "## Introduction\n",
    "For example, download vagina metagenomics data, goto:\n",
    "\n",
    "https://www.ebi.ac.uk/metagenomics/browse?lineage=root:Host-associated:Human:Reproductive%20system:Vagina#studies\n",
    "\n",
    "\n",
    "<img src=\"../images/mgnify_human_vagina_study.png\">\n",
    "\n",
    "When we click: MGY00005262,\n",
    "\n",
    "<img src=\"../images/mgnify_human_vagina_study_2.png\">\n",
    "\n",
    "We can see SRP136173, then download it using mg-toolkit:\n",
    "\n",
    "```bash\n",
    "mg-toolkit original_metadata -a SRP136173\n",
    "```\n",
    "\n",
    "When we click SRP136173, will goto ENA website:\n",
    "\n",
    "<img src=\"../images/mgnify_human_vagina_study_3.png\">\n",
    "\n",
    "In fact, SRP136173 == PRJNA439436, it means, we can also do like below:\n",
    "\n",
    "```bash\n",
    "mg-toolkit original_metadata -a PRJNA439436\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Results\n",
    "\n",
    "### study summary\n",
    "We can download a csv file which describle total vagina study list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 27483    0 27483    0     0   7295      0 --:--:--  0:00:03 --:--:--  7295\n"
     ]
    }
   ],
   "source": [
    "!curl \"https://www.ebi.ac.uk/metagenomics/api/v1/studies?lineage=root%3AHost-associated%3AHuman%3AReproductive+system%3AVagina&ordering=-last_update&format=csv\" > ../sources/mgnify_human_vagina_study.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>url</th>\n",
       "      <th>samples</th>\n",
       "      <th>publications</th>\n",
       "      <th>accession</th>\n",
       "      <th>samples_count</th>\n",
       "      <th>geocoordinates</th>\n",
       "      <th>analyses</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>downloads</th>\n",
       "      <th>biomes</th>\n",
       "      <th>secondary_accession</th>\n",
       "      <th>centre_name</th>\n",
       "      <th>is_public</th>\n",
       "      <th>public_release_date</th>\n",
       "      <th>study_abstract</th>\n",
       "      <th>study_name</th>\n",
       "      <th>data_origination</th>\n",
       "      <th>last_update</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/stud...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGYS00005262</td>\n",
       "      <td>525</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRJNA439436</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{\"type\": \"biomes\", \"id\": \"root:Host-associate...</td>\n",
       "      <td>SRP136173</td>\n",
       "      <td>mothur</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>We conducted a cohort study with crossover des...</td>\n",
       "      <td>The impact of tampon use on the vaginal microb...</td>\n",
       "      <td>HARVESTED</td>\n",
       "      <td>2020-06-21T09:08:05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/stud...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGYS00001070</td>\n",
       "      <td>1902</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRJNA208535</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{\"type\": \"biomes\", \"id\": \"root:Host-associate...</td>\n",
       "      <td>SRP026107</td>\n",
       "      <td>University of Maryland Institue for Genome Sci...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacterial vaginosis (BV) is a condition charac...</td>\n",
       "      <td>The daily dynamics of the vaginal microbiota b...</td>\n",
       "      <td>HARVESTED</td>\n",
       "      <td>2020-06-21T07:18:13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/stud...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGYS00005531</td>\n",
       "      <td>541</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRJEB12500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{\"type\": \"biomes\", \"id\": \"root:Host-associate...</td>\n",
       "      <td>ERP013986</td>\n",
       "      <td>SNU</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>We characterized the vaginal microbiome from 5...</td>\n",
       "      <td>Genetic associaiton with vaginal microbiome in...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2020-06-21T06:51:25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/stud...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGYS00005222</td>\n",
       "      <td>1882</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRJNA489669</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{\"type\": \"biomes\", \"id\": \"root:Host-associate...</td>\n",
       "      <td>SRP159872</td>\n",
       "      <td>Institute for Genome Sciences, University of M...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Amplification, sequencing and analysis of the ...</td>\n",
       "      <td>Ultra-high throughput multiplexing and sequenc...</td>\n",
       "      <td>HARVESTED</td>\n",
       "      <td>2020-04-21T13:59:59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/stud...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGYS00005196</td>\n",
       "      <td>2367</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRJNA393472</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{\"type\": \"biomes\", \"id\": \"root:Host-associate...</td>\n",
       "      <td>SRP115697</td>\n",
       "      <td>Stanford University School of Medicine</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>We conducted a longitudinal study based on wee...</td>\n",
       "      <td>Replication and Refinement of a Vaginal Microb...</td>\n",
       "      <td>HARVESTED</td>\n",
       "      <td>2020-04-12T15:53:05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 url  samples  publications  \\\n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/stud...      NaN           NaN   \n",
       "1  https://www.ebi.ac.uk/metagenomics/api/v1/stud...      NaN           NaN   \n",
       "2  https://www.ebi.ac.uk/metagenomics/api/v1/stud...      NaN           NaN   \n",
       "3  https://www.ebi.ac.uk/metagenomics/api/v1/stud...      NaN           NaN   \n",
       "4  https://www.ebi.ac.uk/metagenomics/api/v1/stud...      NaN           NaN   \n",
       "\n",
       "      accession  samples_count  geocoordinates  analyses   bioproject  \\\n",
       "0  MGYS00005262            525             NaN       NaN  PRJNA439436   \n",
       "1  MGYS00001070           1902             NaN       NaN  PRJNA208535   \n",
       "2  MGYS00005531            541             NaN       NaN   PRJEB12500   \n",
       "3  MGYS00005222           1882             NaN       NaN  PRJNA489669   \n",
       "4  MGYS00005196           2367             NaN       NaN  PRJNA393472   \n",
       "\n",
       "   downloads                                             biomes  \\\n",
       "0        NaN  [{\"type\": \"biomes\", \"id\": \"root:Host-associate...   \n",
       "1        NaN  [{\"type\": \"biomes\", \"id\": \"root:Host-associate...   \n",
       "2        NaN  [{\"type\": \"biomes\", \"id\": \"root:Host-associate...   \n",
       "3        NaN  [{\"type\": \"biomes\", \"id\": \"root:Host-associate...   \n",
       "4        NaN  [{\"type\": \"biomes\", \"id\": \"root:Host-associate...   \n",
       "\n",
       "  secondary_accession                                        centre_name  \\\n",
       "0           SRP136173                                             mothur   \n",
       "1           SRP026107  University of Maryland Institue for Genome Sci...   \n",
       "2           ERP013986                                                SNU   \n",
       "3           SRP159872  Institute for Genome Sciences, University of M...   \n",
       "4           SRP115697             Stanford University School of Medicine   \n",
       "\n",
       "   is_public public_release_date  \\\n",
       "0       True                 NaN   \n",
       "1       True                 NaN   \n",
       "2       True                 NaN   \n",
       "3       True                 NaN   \n",
       "4       True                 NaN   \n",
       "\n",
       "                                      study_abstract  \\\n",
       "0  We conducted a cohort study with crossover des...   \n",
       "1  Bacterial vaginosis (BV) is a condition charac...   \n",
       "2  We characterized the vaginal microbiome from 5...   \n",
       "3  Amplification, sequencing and analysis of the ...   \n",
       "4  We conducted a longitudinal study based on wee...   \n",
       "\n",
       "                                          study_name data_origination  \\\n",
       "0  The impact of tampon use on the vaginal microb...        HARVESTED   \n",
       "1  The daily dynamics of the vaginal microbiota b...        HARVESTED   \n",
       "2  Genetic associaiton with vaginal microbiome in...        SUBMITTED   \n",
       "3  Ultra-high throughput multiplexing and sequenc...        HARVESTED   \n",
       "4  Replication and Refinement of a Vaginal Microb...        HARVESTED   \n",
       "\n",
       "           last_update  \n",
       "0  2020-06-21T09:08:05  \n",
       "1  2020-06-21T07:18:13  \n",
       "2  2020-06-21T06:51:25  \n",
       "3  2020-04-21T13:59:59  \n",
       "4  2020-04-12T15:53:05  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "study_df = pd.read_csv(\"../sources/mgnify_human_vagina_study.csv\")\n",
    "study_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### get each study original data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: mg-toolkit [-h] [-V] [-d]\n",
      "                  {original_metadata,sequence_search,bulk_download} ...\n",
      "\n",
      "Metagenomics toolkit\n",
      "--------------------\n",
      "\n",
      "positional arguments:\n",
      "  {original_metadata,sequence_search,bulk_download}\n",
      "    original_metadata   Download original metadata.\n",
      "    sequence_search     Search non-redundant protein database using HMMER\n",
      "    bulk_download       Download result files in bulks for an entire study.\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  -V, --version         print version information\n",
      "  -d, --debug           print debugging information\n"
     ]
    }
   ],
   "source": [
    "# see how to use mg-toolkit\n",
    "!mg-toolkit -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: mg-toolkit original_metadata [-h] -a ACCESSION [ACCESSION ...]\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  -a ACCESSION [ACCESSION ...], --accession ACCESSION [ACCESSION ...]\n",
      "                        Provide study accession, e.g. PRJEB1787 or ERP001736.\n"
     ]
    }
   ],
   "source": [
    "!mg-toolkit original_metadata -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for bioproject in study_df[\"bioproject\"]:\n",
    "    subprocess.run(f\"env HTTPS_PROXY=http://127.0.0.1:9910 HTTP_PROXY=http://127.0.0.1:9910 mg-toolkit original_metadata -a {bioproject}\", shell=True, check=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
